# IBM Data Science and Machine Learning Capstone Project
# Question 2
# What Areas Should the Agency Focus On?
Should the Department of Housing Preservation and Development of New York City focus on any particular set of boroughs, ZIP codes, or street (where the complaints are severe) for the specific type of complaints you identified in response to Question 1?

In [3]:
#Bucket and Credentials

#import types
import pandas as pd
import numpy as np
from botocore.client import Config
import ibm_boto3
#def __iter__(self): return 0

client_cred = ibm_boto3.client(service_name='s3',
ibm_api_key_id='rA-GmnuRkJ81A9EsDqha63ZmdFzQjxwCEgASTF2F-OJr',
ibm_auth_endpoint='https://iam.cloud.ibm.com/oidc/token',
config=Config(signature_version='oauth'),
endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

bucket = 'ibmdatascienceandmachinelearningc-donotdelete-pr-bkscl1den6u69w'

In [4]:
#Download NYC 311 Calls Dataset
client_cred.download_file(Bucket=bucket, Key='df_raw_cos.pkl', Filename='./df_raw_local.pkl')
df = pd.read_pickle('./df_raw_local.pkl')

# Quiz on Affected Areas

In [21]:
# Question 1: For the complaint types that you selected in the previous module that had at least 80,000 complaints logged, 
#which borough had the highest number of complaints submitted?

S1 = df[df["complaint_type"] == "HEAT/HOT WATER"].groupby("borough")["created_date"].count()
S2 = df[df["complaint_type"] == "HEATING"].groupby("borough")["created_date"].count()
df_aux = pd.DataFrame({"#Hot Water complaints": S1, "#Heating complaints": S2})
display(df_aux)

# Boroughs normalized (custom normalization)
calls_borough = df_aux * len(df_aux) / df_aux["#Hot Water complaints"].sum()
display(calls_borough.sort_values("#Hot Water complaints", ascending=False).head())
print("Boroughs:", len(calls_borough))

# Boroughs normalized
S1_norm = df[df["complaint_type"] == "HEAT/HOT WATER"]["borough"].value_counts(normalize = True)
S2_norm = df[df["complaint_type"] == "HEATING"]["borough"].value_counts(normalize = True)
df_norm = pd.DataFrame({"#Hot Water complaints": S1_norm, "#Heating complaints": S2_norm})
df_norm.round(6)

#Hot Water complaints  #Heating complaints
borough                                                  
BRONX                         431906               195284
BROOKLYN                      399663               190325
MANHATTAN                     298182               137525
QUEENS                        175125                75804
STATEN ISLAND                  12822                 6015
Unspecified                        1               282916

#Hot Water complaints  #Heating complaints
borough                                                  
BRONX                       1.966637             0.889205
BROOKLYN                    1.819822             0.866624
MANHATTAN                   1.357740             0.626205
QUEENS                      0.797413             0.345165
STATEN ISLAND               0.058384             0.027389

Boroughs: 6


#Hot Water complaints  #Heating complaints
BRONX                       0.327773             0.219947
BROOKLYN                    0.303304             0.214362
MANHATTAN                   0.226290             0.154893
QUEENS                      0.132902             0.085377
STATEN ISLAND               0.009731             0.006775
Unspecified                 0.000001             0.318646

In [22]:
# Question2: For the complaint types that you selected in the previous module that had a total number that exceeded 80,000 complaints, 
# which borough had the lowest number of complaints submitted?

S1 = df[df["complaint_type"] == "HEAT/HOT WATER"].groupby("borough")["created_date"].count()
S2 = df[df["complaint_type"] == "HEATING"].groupby("borough")["created_date"].count()
pd.DataFrame({"#Hot Water complaints": S1, "#Heating complaints": S2}).sort_values("#Hot Water complaints")

#Hot Water complaints  #Heating complaints
borough                                                  
Unspecified                        1               282916
STATEN ISLAND                  12822                 6015
QUEENS                        175125                75804
MANHATTAN                     298182               137525
BROOKLYN                      399663               190325
BRONX                         431906               195284

In [23]:
# Question 3: For the complaint types that you selected in the previous module that had at least 80,000 complaints logged,
# which ZIP code had the highest number of complainted submitted?

S1 = df[df["complaint_type"] == "HEAT/HOT WATER"].groupby("incident_zip")["created_date"].count()
S2 = df[df["complaint_type"] == "HEATING"].groupby("incident_zip")["created_date"].count()
display(pd.DataFrame({"#Hot Water complaints": S1, "#Heating complaints": S2}).sort_values("#Hot Water complaints", ascending=False).head())

#zips normalized (custom normalization)
calls_zip = pd.DataFrame({"#Hot Water complaints": S1, "#Heating complaints": S2})
calls_zip = calls_zip * len(calls_zip) / calls_zip["#Hot Water complaints"].sum()
display(calls_zip.sort_values("#Hot Water complaints", ascending=False).head())
print("#Zips:", len(calls_zip))

# Zips normalized
S1_norm = df[df["complaint_type"] == "HEAT/HOT WATER"]["incident_zip"].value_counts(normalize=True)
S2_norm = df[df["complaint_type"] == "HEATING"]["incident_zip"].value_counts(normalize=True)
df_norm = pd.DataFrame({"#Hot Water complaints": S1_norm, "#Heating complaints": S2_norm})
df_norm.sort_values("#Hot Water complaints", ascending=False).head()

#Hot Water complaints  #Heating complaints
incident_zip                                            
11226.0                     43596.0              27255.0
10458.0                     40488.0              26508.0
10467.0                     39757.0              27963.0
10468.0                     36018.0              23683.0
10453.0                     35648.0              23577.0

#Hot Water complaints  #Heating complaints
incident_zip                                            
11226.0                    6.420109             4.013673
10458.0                    5.962413             3.903667
10467.0                    5.854764             4.117935
10468.0                    5.304145             3.487647
10453.0                    5.249657             3.472037

#Zips: 193


#Hot Water complaints  #Heating complaints
11226.0               0.033265             0.031114
10458.0               0.030893             0.030262
10467.0               0.030336             0.031923
10468.0               0.027483             0.027037
10453.0               0.027200             0.026916

In [24]:
# Question 4: For the complaint types that you selected in the previous module that had at least 80,000 complaints logged,
# the address 89-21 Elmhurst Avenue had the highest number of complainted submitted?

S1 = df[df["complaint_type"] == "HEAT/HOT WATER"].groupby("street_name")["created_date"].count()
S2 = df[df["complaint_type"] == "HEATING"].groupby("street_name")["created_date"].count()
display(pd.DataFrame({"#Hot Water complaints": S1, "#Heating complaints": S2}).sort_values("#Hot Water complaints", ascending=False).head())

# Street normalized (custom normalization)
calls_street = pd.DataFrame({"#Hot Water complaints": S1, "#Heating complaints": S2})
calls_street = calls_street * len(calls_street) / calls_street["#Hot Water complaints"].sum()
display(calls_street.sort_values("#Hot Water complaints", ascending=False).head())
print("#Streets:", len(calls_street))

# Street normalized
S1_norm = df[df["complaint_type"] == "HEAT/HOT WATER"]["street_name"].value_counts(normalize=True)
S2_norm = df[df["complaint_type"] == "HEATING"]["street_name"].value_counts(normalize=True)
df_norm = pd.DataFrame({"#Hot Water complaints": S1_norm, "#Heating complaints": S2_norm})
df_norm.sort_values("#Hot Water complaints", ascending=False).head()

#Hot Water complaints  #Heating complaints
GRAND CONCOURSE                23316.0              15576.0
BROADWAY                       15883.0               9116.0
ELMHURST AVENUE                12206.0               1033.0
OCEAN AVENUE                   10609.0               8606.0
MORRIS AVENUE                   9957.0               6682.0

#Hot Water complaints  #Heating complaints
GRAND CONCOURSE             107.157777            71.585587
BROADWAY                     72.996525            41.896136
ELMHURST AVENUE              56.097437             4.747555
OCEAN AVENUE                 48.757800            39.552232
MORRIS AVENUE                45.761279            30.709739

#Streets: 6056


#Hot Water complaints  #Heating complaints
GRAND CONCOURSE               0.017694             0.017543
BROADWAY                      0.012054             0.010267
ELMHURST AVENUE               0.009263             0.001163
OCEAN AVENUE                  0.008051             0.009693
MORRIS AVENUE                 0.007556             0.007526

In [25]:
# Question 5: For the complaint types that you selected in the previous module that had at least 80,000 complaints logged,
# how many of the submitted tickets were closed?

print(df[df["complaint_type"].isin(["HEAT/HOT WATER", "HEATING"])]["closed_date"].notnull().value_counts())

True     2190561
False      15007
Name: closed_date, dtype: int64


# What Areas Should the Agency Focus On?

If analyzing by Borough, it should focus in borough Bronx.<br>
If analyzing by Zip, it should focus in zip 11226.<br>
If analyzing by Street, it should focus in street GRAND CONCOURSE.<br>
<br>
It is sugested that the analysis be done by zip, as it has the highest normalized value.